Closes #14: ADTTE + ADRS with PFS examples (#64)
pharmaverse/examples: adam/adtte.qmd

This example outlines the creation of an ADTTE dataset focusing on Progression-Free Survival using the `admiralonco` package. 

### Required packages
The packages below are required to construct the ADTTE with PFS parameters.
```{r, output=FALSE}
library(admiral)
library(admiralonco)
library(pharmaverseadam)
library(dplyr)
library(lubridate)
```

### Read in Data
Data necessary for the analysis is loaded from the `pharmaverseadam` package.

```{r}
data("adsl")
data("adrs_onco")

adrs <- adrs_onco # renamed for convenience
```

`adsl` contains subject-level information like treatment group, demographic data, etc. `adrs_onco` contains detailed records about adverse events or disease progression specific to oncology. 

### Derive Parameters for Time-to-Event Analysis
This is where the time-to-event parameters are defined. These parameters are used to determine when and how each patient's data is censored or when an event occurs. This can be done using the `derive_param_tte` function from `admiral`.

```{r}
adtte <- derive_param_tte(
    dataset_adsl = adsl,
    start_date = RANDDT,
    event_conditions = list(death_event, pd_event), 

>     censor_conditions = list(lasta_censor, rand_censor), 

Reviewer: you need to first include a section on these source objects and how a user could create their own or else this is a bit mystifying, e.g. see https://pharmaverse.github.io/admiralonco/articles/adtte.html#pre-defined-time-to-event-source-objects

---
title: "ADTTE"
order: 4
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

> This article provides a step-by-step explanation for creating an ADaM `ADTTE` (Time-to-Event) dataset with common oncology endpoint parameters using the `{admiral}`, `{admiralonco}`, `{metacore}`, `{metatools}`, and `{xportr}` packages. ADTTE datasets often involve calculating time-to-event variables for endpoints such as Overall Survival (OS) and Progression-Free Survival (PFS).

Reviewer: ```suggestion
This article provides a step-by-step explanation for creating an ADaM `ADTTE` (Time-to-Event) dataset with common oncology endpoint parameters using key pharmaverse packages along with tidyverse components. `ADTTE` datasets often involve calculating time-to-event variables for endpoints such as Overall Survival (OS) and Progression-Free Survival (PFS).
```
tried to make a bit more consistent with your ADRS intro above

> For the purpose of this example, we will use the `adsl` and `adrs_onco` datasets.

Reviewer: ```suggestion
For the purpose of this example, we will use the `ADSL` and `ADRS_ONCO` datasets from `{pharmaverseadam}`.
```

# Load Required Packages

First, we will load the necessary packages:

```{r message=FALSE, warning=FALSE}
library(admiral)
library(admiralonco)
library(dplyr)
library(lubridate)
library(metacore)
library(metatools)
library(xportr)
library(pharmaverseadam)
```

# Load Specifications and Source Data

> We will load our specification file into a `{metacore}` object to trace the dataset variables and attributes. Then, we will read the source data (`adsl` and `adrs_onco`).

Reviewer: ```suggestion
We will load our specification file into a `{metacore}` object to trace the dataset variables and attributes. Then, we will read the source data.
```

> ```{r, warni}

Reviewer: ```suggestion
```{r read-specs}
#| warning: false
```
In ADRS example you included a label like "read-specs" for each step - would be good to have similar in ADTTE just for consistency.

# Load metacore specifications
metacore <- spec_to_metacore("./metadata/adam_spec.xlsx") %>%
  select_dataset("ADTTE")

# Load source datasets
data("adsl")
data("adrs_onco")

adrs <- adrs_onco %>%

>   convert_blanks_to_na()

Reviewer: `convert_blanks_to_na()` should only be needed for datasets originally created with SAS so these pharmaverseadam datasets should be fine without as they were made with admiral

```

```{r, echo=FALSE}
# Show the first few rows of ADSL and ADRS
head(adsl, n=10)
head(adrs, n=10)
```

# Define Event and Censoring Sources

> We define event and censoring sources using the `admiral::event_source` and `admiral::censor_source` functions. This forms the basis for calculating time-to-event endpoints.

Reviewer: ```suggestion
We define event and censoring sources using the `admiral::event_source()` and `admiral::censor_source()` functions. This forms the basis for calculating time-to-event endpoints.
```

```{r}
# Define event and censoring sources
death_event <- event_source(
  dataset_name = "adsl",
  date = DTHDT,
  set_values_to = exprs(
    EVNTDESC = "Death",
    SRCDOM = "ADSL",
    SRCVAR = "DTHDT"
  )

> )

Reviewer: If you use something like the below instead then it has the added value of linking back to your ADRS example where you created DEATH records so makes the 2 examples work nicer together:

```
death_event <- event_source(
  dataset_name = "adrs",
  filter = PARAMCD == "DEATH" & AVALC == "Y" & ANL01FL == "Y",
  date = ADT,
  set_values_to = exprs(
    EVNTDESC = "Death",
    SRCDOM = "ADRS",
    SRCVAR = "ADT"
  )
)
```

pd_event <- event_source(
  dataset_name = "adrs",
  filter = PARAMCD == "PD" & ANL01FL == "Y",
  date = ADT,
  set_values_to = exprs(
    EVNTDESC = "Progressive Disease",
    SRCDOM = "ADRS",
    SRCVAR = "ADT"
  )
)

lastalive_censor <- censor_source(
  dataset_name = "adsl",
  date = LSTALVDT,
  set_values_to = exprs(
    EVNTDESC = "Last Known Alive",
    CNSDTDSC = "Last Known Alive Date",
    SRCDOM = "ADSL",
    SRCVAR = "LSTALVDT"
  )
)

> 

Reviewer: You're missing a last tumor assessment censor that should be used for PFS such as below. The reason this is used in PFS is you can only truly say a patient didn't have a progression up until the point they last had a tumor assessment where diseases progression would have been checked for. Whereas for OS you only check for death and therefore last alive date is fine to use as a censor.

```
lasta_censor <- censor_source(
  dataset_name = "adrs",
  filter = PARAMCD == "LSTA" & ANL01FL == "Y",
  date = ADT,
  set_values_to = exprs(
    EVNTDESC = "Progression Free Alive",
    CNSDTDSC = "Last Tumor Assessment",
    SRCDOM = "ADRS",
    SRCVAR = "ADT"
  )
)
```

Reviewer: @f-meister you updated this correctly for PFS, but for OS you need to leave the old censor object that David had as:

```
lastalive_censor <- censor_source(
  dataset_name = "adsl",
  date = LSTALVDT,
  set_values_to = exprs(
    EVNTDESC = "Last Known Alive",
    CNSDTDSC = "Last Known Alive Date",
    SRCDOM = "ADSL",
    SRCVAR = "LSTALVDT"
  )
)
```

Then use this as the first censor instead of `lasta_censor` for the OS `derive_param_tte()` call below. The reason is a patient can die at any time, not only at a tumor assessment, so we are able to censor at the last point we know they were alive.

rand_censor <- censor_source(
  dataset_name = "adsl",
  date = RANDDT,
  set_values_to = exprs(
    EVNTDESC = "Randomization Date",
    CNSDTDSC = "Randomization Date",
    SRCDOM = "ADSL",
    SRCVAR = "RANDDT"
  )
)
```

# Derive Time-to-Event Parameters

> The `admiral::derive_param_tte` function is used to derive parameters such as OS (Overall Survival) and PFS (Progression-Free Survival).

Reviewer: ```suggestion
The `admiral::derive_param_tte()` function is used to derive parameters such as OS (Overall Survival) and PFS (Progression-Free Survival).
```

```{r}
# Derive Overall Survival (OS)
adtte <- derive_param_tte(
  dataset_adsl = adsl,
  start_date = RANDDT,
  event_conditions = list(death_event),
  censor_conditions = list(lastalive_censor, rand_censor),
  source_datasets = list(adsl = adsl, adrs = adrs),
  set_values_to = exprs(PARAMCD = "OS", PARAM = "Overall Survival")
)

# Derive Progression-Free Survival (PFS)

> adtte <- adtte %>%

Reviewer: not needed you could just pipe directly from line 120 and add the comment to the next line

  derive_param_tte(
    dataset_adsl = adsl,
    start_date = RANDDT,
    event_conditions = list(pd_event, death_event),

>     censor_conditions = list(lastalive_censor, rand_censor),

Reviewer: ```suggestion
    censor_conditions = list(lasta_censor, rand_censor),
```
See above

    source_datasets = list(adsl = adsl, adrs = adrs),
    set_values_to = exprs(PARAMCD = "PFS", PARAM = "Progression-Free Survival")
  )
```

```{r, echo=FALSE}
# Display first few rows of ADTTE
head(adtte, n=10)
```

# Derive Analysis Value (`AVAL`)

> The analysis value (`AVAL`) can be derived by calling the `admiral::derive_vars_duration` function.

Reviewer: ```suggestion
The analysis value (`AVAL`) can be derived by calling the `admiral::derive_vars_duration()` function.
```

```{r}
# Derive analysis value

> adtte <- adtte %>%

Reviewer: As you did for ADRS good idea to use unique names for each dataframe created as easier for people to test out and debug

  derive_vars_duration(
    new_var = AVAL,
    start_date = STARTDT,
    end_date = ADT
  )
```

```{r, echo=FALSE}
# Display first few rows with AVAL
head(adtte, n=10)
```

# Derive Analysis Sequence Number (`ASEQ`)

> We derive the sequence number for each record to uniquely identify them using the `admiral::derive_var_obs_number` function.

Reviewer: ```suggestion
We derive the sequence number for each record to uniquely identify them using the `admiral::derive_var_obs_number()` function.
```

```{r}
# Derive analysis sequence number
adtte <- adtte %>%
  derive_var_obs_number(
    by_vars = exprs(STUDYID, USUBJID),
    order = exprs(PARAMCD),
    check_type = "error"
  )
```

```{r, echo=FALSE}
# Display first few rows with ASEQ
head(adtte, n=10)
```

# Add ADSL Variables

Additional variables from the `ADSL` dataset are merged into the `ADTTE` dataset using the `admiral::derive_vars_merged` function to enrich it.

```{r}
# Add ADSL variables
adtte <- adtte %>%
  derive_vars_merged(
    dataset_add = adsl,
    by_vars = exprs(STUDYID, USUBJID)
  )
```

```{r, echo=FALSE}
# Display first few rows with merged ADSL variables
head(adtte, n=10)
```

# Apply Metadata and eSub Checks

> We use `{metatools}` and `{xportr}` to perform checks, apply metadata such as types, lengths, labels, and write the dataset to an XPT file.

Reviewer: make this section consistent to how you have it in ADRS and also clean up specs as it currently gives:

Error in `keep()`:
ℹ In index: 1.
Caused by error in `filter()`:
ℹ In argument: `variable %in% names(data)`.
Caused by error in `check_variables()`:
! The following variables are missing:
AAGE
AAGEU
ITTFL
DCUTDT
PARCAT1
PARCAT2
PARCAT3
AVALU
STARTDTF
ADTF
ADY
SRCSEQ
ANL01FL
The following variables do not belong:
RFSTDTC
RFENDTC
RFXSTDTC
RFXENDTC
RFICDTC
RFPENDTC
DTHDTC
DTHFL
ARMCD
ARM
ACTARMCD
ACTARM
DMDTC
DMDY
TRTSTMF
TRTETMF
TRTDURD
SCRFDT
EOSDT
EOSSTT
FRVDT
DTHDT
DTHDTF
DTHADY
LDDTHELD
DTHCAUS
DTHDOM
DTHCGR1
RACEGR1
AGEGR1
LDDTHGR1
DTH30FL
DTHA30FL
DTHB30FL

Reviewer 2: Cleaned up specs, it doesn't show these anymore.
